webhooks/models.py

```diff
@@ -0,0 +1,60 @@
+import hashlib
+import hmac
+import json
+
+from django.conf import settings
+
+
+class GitHookError(Exception):
+    pass
+
+
+class GitHookAccessError(GitHookError):
+    pass
+
+
+REQUIRED_KEYS = ['ref', 'before', 'commits']
+
+
+class GitHook(object):
+
+    def __init__(self, request):
+        self.request = request
+        self.body = None
+
+    def process_request(self):
+        self.validate_header()
+        self.process_body()
+        self.validate_keys()
+        self.generate_properties()
+
+    @property
+    def branch_name(self):
+        return self.ref.replace('refs/heads/', '')
+
+    def validate_header(self):
+        if 'HTTP_X_HUB_SIGNATURE' not in self.request.META:
+            raise GitHookAccessError('Wrong signature')
+
+        sig = self.request.META['HTTP_X_HUB_SIGNATURE']
+        text = self.request.body
+
+        secret = str.encode(settings.WEBHOOK_SECRET)
+        signature = 'sha1=' + hmac.new(secret, msg=text, digestmod=hashlib.sha1).hexdigest()
+
+        if not hmac.compare_digest(sig, signature):
+            raise GitHookAccessError('Signature does not match')
+
+    def validate_keys(self):
+        if not all(key in self.body for key in REQUIRED_KEYS):
+            raise GitHookError("Incomplete data")
```

Comment:
If a key is missing, it's likely that the type of webhook we're receiving isn't one we care about (a deleted branch, for example, is still a "PUSH" webhook, but not what we want). We shouldn't reply with a 400, because GitHub might stop sending webhooks. Instead, we should silently ignore those webhooks.